### DISCUSSION TITLE
FIREBASE_UI - REWRITE of the TableBuilder

### DISCUSSION THREAD

**Discussion by wer-mathurin** [firebase/flutterfire#8569] [Feature request]
@lesnitsky @rrousselGit 
Is there a process to propose a complete REWRITE of the TableBuilder.
I can provide the draft implementation.  The objective it to enable easy customization and have a default implementation that support sorting.  It will be more align with the other widget FirestoreListView

Here is an example how we would be able to use a new  FirestorePaginatedTableBuilder :

```
class SomePage extends StatelessWidget {
  const SomePage ({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FirestorePaginatedTableBuilder<SomeEntity>(
      query: casesCollection,
      columns: [
        SortableDataColumn(
          label: const Text('Column #1'),
          comparable: (d) => d.id,
        ),
        SortableDataColumn(
          label: const Text('Column #2'),
          numeric: true,
          comparable: (d) => d.amount ?? 0,
        ),
        SortableDataColumn(
          label: const Text('Column #3'),
        ),
      ],
      datasource: _CaseDataSource(),
    );
  }
}

class _CaseDataSource extends FirebaseDataSource<SomeEntity> {
  _CaseDataSource();

  @override
  List<DataCell> cells(DocumentReference<SomeEntity>? doc, SomeEntity item) {
    return [
      DataCell(
        Text(item.id),
        onTap: () async {
          final copy = item.copyWith(id: 'SOMETHING');
          doc?.set(copy, SetOptions(merge: true));
        },
      ),
      DataCell(Text('$item.amount')),
      DataCell(Text('${item.opened}')),
    ];
  }
}
```

**Comment by lesnitsky** [comment]
I think the first good step is to outline the problems of the existing implementation.

> The objective it to enable easy customization

What is hard to customize with existing implementation?

> and have a default implementation that support sorting

Are there ways to add this functionality to the existing code instead of rewriting it?

@rrousselGit is an original author of the `TableBuilder`, so could give more input on this

**Reply by wer-mathurin** [reply]
When I said complete rewrite, is because this is a complete new widget that do not comply with the current API expose by .  But this is not a big deal, since this widget is "hidden"...not mention in the documentation :-)
We can provide a complete new Widget to not break the current implementation !

The current implementation is relying on the <String, Objects?> for row instead of relying on an actual Class implementation.
So we are not leveraging the withConverter that we can apply on collection.

We will be align with other widget like this:
```
/// FirestoreListView<Movie>(
///   query: moviesCollection,
///   itemBuilder: (context, snapshot) {
///     Movie movie = snapshot.data();
///     return Text(movie.title);
///   },
/// )
```

What is hard to customize with existing implementation?

- Sorting : You can provide the initial sorting and there is no way to change that after...so this is not hard, this is just not possible with the current implementation 
- Reordering column: This is not supported by the underlying DataTable (https://github.com/flutter/flutter/issues/85479)
I can eventually look at an implementation at my SortablePaginatedDataTableBuilder<T> this will be easier to do since, we will Cells will be draw base on col key (https://github.com/firebase/flutterfire/pull/8539/commits/edb6081810cf9977b4b8428b36b8cbb229a0e349)
 

I can publish my code my github repo, so you can look at my implementation, if you wish.

**Comment by rrousselGit** [comment]
I'm not against a rewrite, but we need issues talking about the problem first.

Could you make one issue per problem with the current implementation? So that we can discuss about them individually

Otherwise it's difficult to move forward as that proposal is a bit too broad at the moment.
